### PR TITLE
Fix 14927

### LINKED ISSUE #14927: devtool source-map creates broken `webpack://` urls when `output.library.name` is an array containing a placeholder
# Bug report

**What is the current behavior?**

When using an `output.library.name` containing a `[name]` placeholder, e.g. `['MyLib', '[name]']`, the source maps generated by `'devtool': 'source-map'` are broken. I get a bunch of JS console warnings like this when using my scripts:

```
Source map error: Error: Invalid URL: webpack://MyLib.[name]/webpack/bootstrap
```
image

The problem seems to be that the `[name]` placeholder in `output.library.name` is being treated as literal in the source map URLs, rather than being filled in with the entry file name. Source maps aside, the JS does work as expected, with the placeholder being replaced in other contexts. As far as I can tell, this is a bug, as using the `[name]` placeholder in `output.library.name` is a [supported feature](https://github.com/webpack/webpack/commit/9e6b6deea4076322ae68d122c19e99bd4b9d162b).

**If the current behavior is a bug, please provide the steps to reproduce.**

1. Create an empty folder. Add the following files:

`webpack.config.js`:
```js
const path = require('path')

module.exports = {
	mode: 'production',
	context: path.resolve(__dirname, 'src'),
	entry: { foo: './foo.js' },
	output: {
		path: path.resolve(__dirname, 'build'),
		filename: '[name].js',
		library: { name: ['MyLib', '[name]'], type: 'window' },
		clean: true,
	},
	devtool: 'source-map',
	module: {
		rules: [{ test: /\.js$/u, use: ['babel-loader'] }],
	},
	cache: true,
}
```
`package.json`:
```json
{
	"browserslist": "last 5 Chrome versions, last 3 Firefox versions, last 3 Safari versions, last 3 ChromeAndroid versions, last 2 Android versions, last 2 FirefoxAndroid versions",
	"scripts": {
		"build": "webpack"
	},
	"devDependencies": {
		"@babel/core": "^7.16.0",
		"@babel/preset-env": "^7.16.4",
		"@babel/preset-react": "^7.16.0",
		"babel-loader": "^8.2.3",
		"prettier": "^2.5.1",
		"webpack": "^5.65.0",
		"webpack-cli": "^4.9.1"
	}
}
```
`babel.config.json`:
```json
{ "presets": [["@babel/preset-env", { "spec": true }]] }
```
`demo.html`:
```html
<html>
	<head>
		<meta charset="utf-8" />
		<script defer src="./build/foo.js"></script>
	</head>
	<body>Look at the JS dev console.</body>
</html>
```
`src/foo.js`:
```js
export function myThing() {
	console.log('beep')
}
```

2. Run `npm i && npm run build`.
3. Open `demo.html`.
4. Check the JS dev console.

**What is the expected behavior?**

The source maps should contain working `webpack://` URLs with the `[name]` placeholder filled in as expected.

**Other relevant information:**
webpack version:  5.65.0
Node.js version: 16.13.1
Operating System: KDE Neon Linux

### PR DESCRIPTION
**Pull request by TrickyPi**
**What kind of change does this PR introduce?**

Closes #14927

**Did you add tests for your changes?**

yes

**Does this PR introduce a breaking change?**

no

**What needs to be documented once your changes are merged?**

nothing

### CODE REVIEW

**Comment by ZebulanStanphill** [comment]
This seems to fix the issue and work as expected, but some of the [automated tests](https://dev.azure.com/webpack/webpack/_build/results?buildId=15532&view=results) are failing.

I think the first one is a case where the [snapshot](https://github.com/webpack/webpack/blob/main/test/__snapshots__/StatsTestCases.basictest.js.snap) just needs to be updated, but I could be wrong.
```
FAIL test/StatsTestCases.basictest.js (16.894 s, 89 MB heap size)
  ● StatsTestCases › should print correct stats for related-assets

    expect(received).toMatchSnapshot()
        assets by path *.css 144 bytes
          asset exclude1-chunk_js.css 74 bytes [emitted]
            hidden assets 148 bytes 2 assets
    @@ -67,11 +67,11 @@
          asset exclude2-main.js 14.5 KiB [emitted] (name: main)
            hidden assets 12.5 KiB 1 asset
            compressed exclude2-main.js.br 14.5 KiB [emitted]
            compressed exclude2-main.js.gz 14.5 KiB [emitted]
          asset exclude2-chunk_js.js 804 bytes [emitted]
    -       hidden assets 295 bytes 1 asset
    +       hidden assets 298 bytes 1 asset
            compressed exclude2-chunk_js.js.br 804 bytes [emitted]
            compressed exclude2-chunk_js.js.gz 804 bytes [emitted]
        assets by path *.css 144 bytes
          asset exclude2-chunk_js.css 74 bytes [emitted]
            hidden assets 197 bytes 1 asset

      192 | 					.replace(/(\w)\\(\w)/g, "$1/$2")
      193 | 					.replace(/, additional resolving: X ms/g, "");
    > 194 | 				expect(actual).toMatchSnapshot();
          | 				               ^
      195 | 				if (testConfig.validate) testConfig.validate(stats, stderr.toString());
      196 | 				done();
      197 | 			});

      at test/StatsTestCases.basictest.js:194:20
      at lib/MultiCompiler.js:554:14
      at processQueueWorker (lib/MultiCompiler.js:491:6)

 › 1 snapshot failed.
```

The second one is just formatting. You need to use tabs instead of spaces in the test files.
```
/home/vsts/work/1/s/test/configCases/source-map/source-map-in-multi-entry/webpack.config.js
   3:1  error  Replace `····` with `↹`       prettier/prettier
   4:1  error  Replace `····` with `↹`       prettier/prettier
   5:1  error  Replace `········` with `↹↹`  prettier/prettier
   6:1  error  Replace `········` with `↹↹`  prettier/prettier
   7:1  error  Replace `····` with `↹`       prettier/prettier
   8:1  error  Replace `····` with `↹`       prettier/prettier
   9:1  error  Replace `········` with `↹↹`  prettier/prettier
  10:1  error  Replace `········` with `↹↹`  prettier/prettier
  11:1  error  Replace `····` with `↹`       prettier/prettier
```

**Comment by TrickyPi** [comment]
I agree with you, I will update the snapshot later today

File: lib/SourceMapDevToolPlugin.js [hunk 1]
```diff
@@ -283,17 +283,22 @@ class SourceMapDevToolPlugin {
 
 								if (task) {
 									const modules = task.modules;
-
+									const chunk = fileToChunk.get(file);
+									const { name } = (chunk && chunk.getEntryOptions()) || {
+										name: "*"
+									};
+									const namespaceEnsure = namespace.replace(/\[name\]/g, name);
```
Comment (ZebulanStanphill) [inline lib/SourceMapDevToolPlugin.js:290]: I think a clearer name for this variable would be `parsedNamespace`.

File: lib/SourceMapDevToolPlugin.js [hunk 2]
```diff
@@ -283,17 +283,22 @@ class SourceMapDevToolPlugin {
...
 									for (let idx = 0; idx < modules.length; idx++) {
 										const module = modules[idx];
+										const moduleUnique = module + "|" + name;
```
Comment (ZebulanStanphill) [inline lib/SourceMapDevToolPlugin.js:293]: I think a clearer name for this variable would be `moduleUniqueKey`.

**Review by ZebulanStanphill — APPROVED** [review]
I have some variable renaming suggestions, but aside from that, this looks good to merge, assuming tests pass.

File: lib/SourceMapDevToolPlugin.js (hunk 1 above)
Comment (TrickyPi) [reply lib/SourceMapDevToolPlugin.js:290]: That's a great suggestion

File: lib/SourceMapDevToolPlugin.js (hunk 2 above)
Comment (TrickyPi) [reply lib/SourceMapDevToolPlugin.js:293]: That's a great suggestion

File: lib/SourceMapDevToolPlugin.js (hunk 1 above)
Comment (TrickyPi) [reply lib/SourceMapDevToolPlugin.js]: Done

File: lib/SourceMapDevToolPlugin.js (hunk 2 above)
Comment (TrickyPi) [reply lib/SourceMapDevToolPlugin.js]: Done

**Comment by TrickyPi** [comment]
The `codecov/changes/integration` automated test failed. Because i added a new test folder?

**Comment by alexander-akait** [comment]
Ignore this

**Review by TrickyPi — COMMENTED** [review]
Done

**Review by sokra — CHANGES_REQUESTED** [review]
Please see https://github.com/webpack/webpack/issues/14927#issuecomment-1013066156

File: lib/config/defaults.js [hunk 3]
```diff
@@ -704,8 +704,11 @@ const applyOutputDefaults = (
 			"type" in library
 				? library.name
 				: /** @type {LibraryName=} */ (library);
+		const excludePlaceholders = ["[name]"];
 		if (Array.isArray(libraryName)) {
-			return libraryName.join(".");
+			return libraryName
+				.filter(item => !excludePlaceholders.includes(item))
```
Comment (vankop) [inline lib/config/defaults.js:710]: you probably want to filter any placeholder

Comment (TrickyPi) [reply lib/config/defaults.js:710]: Some placeholders are invalid in `library` configuration, eg [path].

Comment (TrickyPi) [reply lib/config/defaults.js:710]: Could we only filter some placeholders that are valid for `library` configuration?

Comment (vankop) [reply lib/config/defaults.js:710]: I think you can use regexp `/^\[.*\]$/`

Comment (TrickyPi) [reply lib/config/defaults.js:710]: an edge case, if someone want to use `[invalidPlaceholder]` as the string. The regexp `/^\[.*\]$/` will filter it.

Comment (TrickyPi) [reply lib/config/defaults.js:710]: Hmm, i think the edge case above is weird, maybe we don't worry about it?

Comment (alexander-akait) [reply lib/config/defaults.js:710]: We can add this to docs and avoid extra checks

Comment (TrickyPi) [reply lib/config/defaults.js:710]: Well, the pr can be closed.

**Comment by TrickyPi** [comment]
Thanks for your help and review.

**Comment by sokra** [comment]
Thanks

**Pull request merged**